### PR TITLE
Create the skeleton of a Django project

### PR DESCRIPTION
**Pull request by igorkruglyak**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -43,11 +43,22 @@ repos:
   hooks:
   - id: mypy
 
-- repo: local
+- repo: git://github.com/pre-commit/mirrors-pylint.git
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: Use local.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -43,11 +43,22 @@ repos:
...
-    args: []
+    args:
+    - --msg-template={path}:L {line:3d}:({symbol}) {msg} (C {column:2d}), ::{msg_id}
+    - --output-format=colorized
+    - |-
+        --init-hook=
+        import os, platform, sys;
+        sys.path.insert(
+            0,
+            "".join([os.getenv("HOME"), "/virtualenv/python", platform.python_version(),
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: Don't use this hack.

File: mysite/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+"""This is the example module.
```
Comment (webknjaz) [inline mysite/__init__.py]: This line is bullshit.

File: mysite/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+
+This module does stuff.
```
Comment (webknjaz) [inline mysite/__init__.py]: And this too.

File: mysite/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+"""
```
Comment (webknjaz) [inline mysite/__init__.py]: Putting useless stuff in a docstring is not a solution. Go read PEP 257.

File: mysite/__init__.py (hunk 4 above)
Comment (webknjaz) [inline mysite/__init__.py]: And also it's not a module but a package docstring.

File: mysite/settings.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import os
+
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/2.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
```
Comment (webknjaz) [inline mysite/settings.py]: Just keep this in mind for the future.

File: mysite/urls.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+"""mysite URL Configuration.
```
Comment (webknjaz) [inline mysite/urls.py]: Don't keep this whole boilerplate here. Use an actual description.

File: mysite/wsgi.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+"""
+WSGI config for mysite project.
+
+It exposes the WSGI callable as a module-level variable named ``application``.
```
Comment (webknjaz) [inline mysite/wsgi.py]: This sounds like a Wikipedia explanation. Common things shouldn't be explained.

File: requirements.txt [hunk 9]
```diff
@@ -1 +1,3 @@
 Django~=2.2.2
+pylint==2.3.1
```
Comment (webknjaz) [inline requirements.txt]: These aren't runtime deps. Create a separate `test-requirements.txt` file for them. They are only used for testing.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 2 above)
Comment (igorkruglyak) [reply .pre-commit-config.yaml]: I do not clear understand. What do I have to do?

File: requirements.txt (hunk 9 above)
Comment (igorkruglyak) [reply requirements.txt]: I have to create a file test-requirements.txt 
and put it there
Django~=2.2.2
pylint==2.3.1 
Right?

File: mysite/urls.py (hunk 7 above)
Comment (igorkruglyak) [reply mysite/urls.py]: What should I write?

File: .pre-commit-config.yaml (hunk 2 above)
Comment (webknjaz) [reply .pre-commit-config.yaml]: Revert this change. Unfortunately, you cannot do this via `git revert` because it's not in a separate atomic commit. So you'd have to do smth like `git checkout HEAD -- .pre-commit-config.yaml`. Use atomic commits to make your life easier.

File: mysite/urls.py (hunk 7 above)
Comment (webknjaz) [reply mysite/urls.py]: Something that explains what the module is for, not django usage instructions.

File: requirements.txt (hunk 9 above)
Comment (webknjaz) [reply requirements.txt]: No. Don't put django there. Only add test-related things.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (serhii73) [reply .pre-commit-config.yaml]: Why local better than git://github.com/pre-commit/mirrors-pylint.git ?

Comment (webknjaz) [reply .pre-commit-config.yaml]: Because it was the actual task to do so. And also because you shouldn't use hacks.

File: requirements.txt (hunk 9 above)
Comment (serhii73) [reply requirements.txt]: Why use `~=` with Django is a good idea, but use `pylint~=2.3.1` is not a good idea? @webknjaz

Comment (webknjaz) [reply requirements.txt]: It is not about the version pinning. It's about mangling runtime requirements with test requirements.

File: test-requirements.txt [hunk 10]
```diff
@@ -0,0 +1,3 @@
+pre-commit==1.17.0
```
Comment (webknjaz) [inline test-requirements.txt:2]: Now, the runtime requirements are also needed. Just don't copy-paste things. You can just refer to the other file like this:
```suggestion
-r requirements.txt
pre-commit==1.17.0
```

File: requirements.txt [hunk 11]
```diff
@@ -1 +1 @@
-Django~=2.2.2
+Django==2.2.2
```
Comment (webknjaz) [inline requirements.txt]: Plz don't change this now, it's unrelated.

File: .travis.yml [hunk 12]
```diff
@@ -2,5 +2,6 @@ dist: xenial
 language: python
 python: 3.7
 cache: pip
-before_install: pip install -U pre-commit
+before_install:
```
Comment (webknjaz) [inline .travis.yml]: Plz split commits to have testing-related changes separately. Atomic changes are preferrable.

File: requirements.txt (hunk 9 above)
Comment (serhii73) [reply requirements.txt]: Sorry, @webknjaz I don't understand why need `~=` if exist [dependabot](https://dependabot.com).

Comment (webknjaz) [reply requirements.txt]: Don't confuse lib/dev deps (open ended) and app/test/env deps (pinned).

File: mysite/wsgi.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+"""WSGI config for mysite project."""
+
+
+# pylint: disable=invalid-name
```
Comment (webknjaz) [inline mysite/wsgi.py]: This is the wrong place. `os` name is not invalid. Put it where the linter complaint exists.

File: test-requirements.txt (hunk 10 above)
Comment (webknjaz) [reply test-requirements.txt:2]: This is still unsolved.

File: mysite/urls.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+"""URL conf for mysite."""
+from django.contrib import admin
+from django.urls import path
+
+# pylint: disable=invalid-name
```
Comment (webknjaz) [inline mysite/urls.py]: This is a django-specific name which is required by the framework. It's going to be used in other `urls.py` files. So it makes sense to add it to the global list of good names (unless that pylint-django plugin already handles this case).

File: mysite/urls.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
+"""URL conf for mysite."""
+from django.contrib import admin
+from django.urls import path
+
```
Comment (webknjaz) [inline mysite/urls.py:4]: I'm pretty sure that two empty lines needed to separate imports from vars.

File: mysite/wsgi.py (hunk 13 above)
Comment (webknjaz) [inline mysite/wsgi.py]: This should be put onto the offensive line.

File: mysite/urls.py (hunk 14 above)
Comment (webknjaz) [inline mysite/urls.py]: This is a django-specific name which is required by the framework. It's going to be used in other urls.py files. So it makes sense to add it to the global list of good names (unless that pylint-django plugin already handles this case).

File: .gitignore [hunk 16]
```diff
@@ -183,10 +183,6 @@ tags
 
 ### VisualStudioCode ###
 .vscode/*
-!.vscode/settings.json
```
Comment (webknjaz) [inline .gitignore]: Why is this related?

File: mysite/__init__.py [hunk 17]
```diff
@@ -0,0 +1 @@
+"""Initialize the mysite package."""
```
Comment (webknjaz) [inline mysite/__init__.py]: It's a module, not an action. `Initialize` is a verb implying that this module does something once imported which is a lie.

File: mysite/settings.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
+"""
+Django settings for mysite project.
+
+Generated by 'django-admin startproject' using Django 2.2.2.
```
Comment (webknjaz) [inline mysite/settings.py]: This is not needed.

File: mysite/wsgi.py [hunk 19]
```diff
@@ -0,0 +1,10 @@
+"""WSGI config for mysite project."""
+
+
+import os
+
+from django.core.wsgi import get_wsgi_application
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mysite.settings')
+# pylint: disable=invalid-name
+application = get_wsgi_application()
```
Comment (webknjaz) [inline mysite/wsgi.py]: It is better to put comments in the same line if they fit the line length limit.

File: .gitignore (hunk 16 above)
Comment (igorkruglyak) [reply .gitignore]: Because I think these exceptions are not needed on the server

Comment (webknjaz) [reply .gitignore]: Doesn't sound related to adding a skeleton. It'll still work without this. So it doesn't belong to this PR.

**Comment by webknjaz** [comment]
> Fix comments

Invent something better. What kind of comments?

File: .gitignore (hunk 16 above)
Comment (igorkruglyak) [reply .gitignore]: Should I make squash with the previous commit?

Comment (webknjaz) [reply .gitignore]: It doesn't belong to the PR, nor to that commit. So no. Just move it out.

Comment (igorkruglyak) [reply .gitignore]: rebase -i
delete commit?

Comment (webknjaz) [reply .gitignore]: You can cherry-pick it to a separate branch if you still want to apply it. Just don't keep it in the branch bound to this PR.

Comment (serhii73) [reply .gitignore]: I think not need to change gitignote in this pr. If you want to change gitignote, need create a new pr

Comment (igorkruglyak) [reply .gitignore]: thanks

File: .pre-commit-config.yaml [hunk 20]
```diff
@@ -50,4 +50,5 @@ repos:
     name: PyLint
     files: \.py$
     entry: python3 -m pylint.__main__
-    args: []
+    args:
+    - --good-names=pk,urlpatterns
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: Why `pk`? Where is it used?

File: mysite/__init__.py [hunk 21]
```diff
@@ -0,0 +1 @@
+"""Initialization module."""
```
Comment (webknjaz) [inline mysite/__init__.py]: Nope. The docstring in `__init__` should about the `mysite` package. Also, `Initialization module` doesn't actually mean anything useful.
Imagine you open this file and expect to see some hint on what's inside `mysite/` — will this sentence help you?

File: mysite/wsgi.py [hunk 22]
```diff
@@ -0,0 +1,10 @@
+"""WSGI config for mysite project."""
+
+
```
Comment (webknjaz) [inline mysite/wsgi.py]: You don't really need this many spaces before the imports section. Look at the style in other modules. PEP8 encourages us to be consistent wrt style.

File: mysite/__init__.py (hunk 21 above)
Comment (igorkruglyak) [reply mysite/__init__.py]: Initializes the mysite directory as a standard python package? So it will be right?

File: .pre-commit-config.yaml (hunk 20 above)
Comment (igorkruglyak) [reply .pre-commit-config.yaml]: No, I will not use it anywhere. Sorry i will correct it.

File: mysite/__init__.py (hunk 21 above)
Comment (webknjaz) [reply mysite/__init__.py]: No.
1) You don't need to explain how Python works. It's not a Wikipedia article.
2) "Initializes" is a verb. It's a module. A declaration. NOT an action. It does not DO anything, it IS something.

File: mysite/wsgi.py (hunk 22 above)
Comment (igorkruglyak) [reply mysite/wsgi.py]: Ok.

File: mysite/__init__.py (hunk 21 above)
Comment (igorkruglyak) [reply mysite/__init__.py]: What should I write?

Comment (webknjaz) [reply mysite/__init__.py]: Something like `Django website package.`

But the ultimate goal of this all going back-and-forth with choosing docstrings is to make you understand the ground rules and learn to make useful docstrings.

Comment (igorkruglyak) [reply mysite/__init__.py]: Ok. Thanks.

File: mysite/wsgi.py (hunk 22 above)
Comment (serhii73) [reply mysite/wsgi.py]: https://github.com/PyCQA/pydocstyle/ - I think it is a good tool for doc check.

Comment (webknjaz) [reply mysite/wsgi.py]: @serhii73 it's already integrated: https://github.com/kpi-web-guild/django-girls-blog-igorkruglyak/blob/e7ee46c/.pre-commit-config.yaml#L39.
But there's things which automatic tooling just incapable of checking. That's why we need a human review.

**Comment by webknjaz** [comment]
@igorkruglyak you've just added a merge commit. I'm going to ask you to rebase instead.

**Comment by webknjaz** [comment]
Let's review the commit messages (https://github.com/kpi-web-guild/django-girls-blog-igorkruglyak/pull/22/commits):

> Create the skeleton of a Django project

:+1: 

> Add test-requirements.txt fix requirements.txt

:-1: 

> Fix wsgi.py and test-requirements.txt

:-1: 

> Add good-names in pre-commit

:+1: :-1: AFAIK you're adding the setting of pylint here

> Write doc for the package

docstrings, not a doc. we don't learn to write docs at this point.


So things need to be fixed here.

**Comment by igorkruglyak** [comment]
> Let's review the commit messages (https://github.com/kpi-web-guild/django-girls-blog-igorkruglyak/pull/22/commits):
> 
> > Create the skeleton of a Django project
> 
> +1
> 
> > Add test-requirements.txt fix requirements.txt
> 
> -1
> 
> > Fix wsgi.py and test-requirements.txt
> 
> -1
> 
> > Add good-names in pre-commit
> 
> +1 -1 AFAIK you're adding the setting of pylint here
> 
> > Write doc for the package
> 
> docstrings, not a doc. we don't learn to write docs at this point.
> 
> So things need to be fixed here.

Maybe change
```Add test-requirements.txt fix requirements.txt``` on 
```Separation dependency files ```

```Fix wsgi.py and test-requirements.txt```
``` Fix settings in wsgi.py and test-requirements.txt```

```Add good-names in pre-commit```
```Add pylint settings```

```Write doc for the package```
 ```Write docstrings for the package```

**Comment by webknjaz** [comment]
> Separation dependency files

This should start with a verb.

**Comment by webknjaz** [comment]
> Fix settings in wsgi.py and test-requirements.txt

Why does this state two separate actions? And what settings did you fix in the test-requirements?

**Comment by webknjaz** [comment]
> Add pylint settings

Too generic. What exactly did you do with pylint settings?

**Comment by webknjaz** [comment]
> Write docstrings for the package

I just took a look at the commit and now I see that it contains a lot of weird changes like adding an empty line or changing pylint settings. This is wrong. The commit should be atomic.

Also, adding docstrings could be squashed into the initial commit.

**Comment by webknjaz** [comment]
rebase needed

File: test-requirements.txt [hunk 23]
```diff
@@ -1,2 +1,3 @@
+-r requirements.txt
 pre-commit==1.17.0
 pylint==2.3.1
```
Comment (webknjaz) [inline test-requirements.txt:3]: It looks like you've lost that pylint plugin during rebase after all

File: manage.py [hunk 24]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python
+"""Django's command-line utility for administrative tasks."""
+import os
+import sys
+
+
+def main():
+    """Blog app manager."""
```
Comment (webknjaz) [inline manage.py]: It's a function. Use a verb in the imperative mood.
```suggestion
    """Invoke blog app manager."""
```

File: mysite/__init__.py [hunk 25]
```diff
@@ -0,0 +1 @@
+"""Django website package."""
```
Comment (webknjaz) [inline mysite/__init__.py]:
```suggestion
"""Django website."""
```

Comment (webknjaz) [reply mysite/__init__.py]: Still unsolved.

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Comment by webknjaz** [comment]
> Add pylint settings to wsgi.py

This one is odd: it does nothing related to the commit message.

**Review by webknjaz — APPROVED** [review]

**Comment by serhii73** [comment]
@webknjaz 
Maybe https://github.com/SvanBoxel/delete-merged-branch will be useful for you.
This tool can automatically remove merge branches.

**Comment by webknjaz** [comment]
It's not built into project settings. No need for a separate bot.
